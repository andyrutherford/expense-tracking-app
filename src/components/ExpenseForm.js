import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            category: props.expense ? props.expense.category : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onCategoryChange = (e) => {
        console.log('category is changing!');
        const category = e.target.value;
        this.setState(() => ({ category }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        //Enforce number is of format XXX.XX
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(()=> ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        //Don't allow user to clear date for date picker
        if(createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        //Prevent page refresh on submit
        e.preventDefault();
        if(!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide a description and amount.' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                category: this.state.category,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit} >
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input 
                    text="text"
                    placeholder="Description"
                    autoFocus
                    className="text-input"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input 
                    type="text"
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <select
                    className="select"
                    value={this.state.category}
                    onChange={this.onCategoryChange}
                >
                    <option 
                        value="none">Category</option>
                    <option 
                        value="personal">Personal</option>
                    <option 
                        value="family">Family</option>
                    <option 
                        value="business">Business</option>
                </select>
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false }
                />
                <textarea
                    placeholder="Add a note for your expense"
                    className="textarea"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        );
    }
}