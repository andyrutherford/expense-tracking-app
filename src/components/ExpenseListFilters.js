import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByCategory, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null,
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };
    onCategoryChange = (e) => {
        const category = e.target.value;
        this.props.sortByCategory(category);
    };
    render() {
        return (
            <div>
                <input 
                    type="text" 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}
                />
                <select
                    className="sort-change-cat"
                    value={this.props.filters.sortByCategory}
                    onChange={this.onCategoryChange}
                >
                <option 
                    value="any">All Categories</option>
                <option 
                value="personal">Personal</option>
                <option 
                    value="family">Family</option>
                <option 
                    value="business">Business</option>
                </select>
                <select
                    className="sort-change"
                    value={this.props.filters.sortBy}
                    onChange={this.onSortChange}
                >
                    <option 
                        value="date">Date</option>
                    <option 
                        value="amount">Amount</option>
                </select>
                <DateRangePicker 
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByCategory: (category) => dispatch(sortByCategory(category)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);