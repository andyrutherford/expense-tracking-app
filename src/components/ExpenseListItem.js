import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';


const ExpenseListItem = ( { description, category, amount, createdAt, id } ) => (
    <div>
        <Link 
            to={`edit/${id}`}
            ><h3>{description}</h3>
        </Link>
        {category && (<p>[{category}]</p>) }
        <p>
            {numeral(amount / 100).format('$0,0.00')}{' '}
            - 
            {' '}{moment(createdAt).format('DD MMMM YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;

