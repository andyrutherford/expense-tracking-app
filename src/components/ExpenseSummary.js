import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import totalExpenses from '../selectors/expenses-total';


export const ExpenseSummary = ({ expenseCount, expensesTotal }) => (
    <div>
        {expenseCount === 0 ? (
            <p>No expenses to display.</p>
        ) : (
            <p>Showing {expenseCount} {expenseCount === 1 ? ("expense") : ("expenses")} totalling {numeral(expensesTotal / 100).format('$0,0.00')}.</p>
        ) }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenseCount: selectExpenses(state.expenses, state.filters).length,
        expensesTotal: totalExpenses(selectExpenses(state.expenses, state.filters))
    };
};

export default connect(mapStateToProps)(ExpenseSummary);