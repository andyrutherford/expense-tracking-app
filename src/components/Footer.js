import React from 'react';

const Footer = () => {
    return (
        <footer className="footer">
        <div className="content-container">
            <p>Made with &lt;3 by <a href="https://www.andyrutherford.com" target="_blank">Andy Rutherford</a></p>
        </div>
    </footer>
    );
};

export default Footer;

