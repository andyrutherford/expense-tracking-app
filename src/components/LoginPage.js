import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = ( { startLogin } ) => (
    <div className="box-layout">
        <div className="box-layout__box">
            <h1 className="box-layout__title">Expense Tracking App</h1>
            <p>Take control of your expenses.</p>
            <button
                className="button"
                onClick={startLogin}
            >Login with Google &nbsp;<img className="button__icon" src="images/google.png" /></button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);