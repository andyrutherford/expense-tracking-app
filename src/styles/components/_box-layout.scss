.box-layout {
    align-items: center;
    background:     
        linear-gradient(
         rgba(0, 0, 0, 0.5), 
         rgba(0, 0, 0, 0.5)
        ), url('/images/bg.jpg');
    background-size: cover;
    display: flex;
    height: 100vh;
    justify-content: center;
    width: 100vw;
}

.box-layout__box {
    background: fade-out(white, .15);
    border-radius: 5px;
    padding: $l-size $m-size;
    text-align: center;
    width: 25rem;
}

.box-layout__title {
    font-size: $font-size-xl; 
    font-family: $logo-font;
    font-weight: 100;
    margin: 0 0 $m-size 0;
    line-height: 1;
}