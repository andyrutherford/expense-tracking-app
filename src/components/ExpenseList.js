import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    {props.expenses.length > 0 && (
        props.expenses.map((expense, index) => (
            <ExpenseListItem 
                key={expense.id}
                {...expense} />
        ))
    )}
  </div>  
);

//Allow ExpenseList to have access to redux store and pass in as props

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);
