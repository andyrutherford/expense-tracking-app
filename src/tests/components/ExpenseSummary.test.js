import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';

test('should render ExpenseSummary with 1 expense', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={1} expensesTotal={56940} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseSummary with multiple expenses', () => {
    const wrapper = shallow(<ExpenseSummary expenseCount={5} expensesTotal={5438757498} />);
    expect(wrapper).toMatchSnapshot();
});