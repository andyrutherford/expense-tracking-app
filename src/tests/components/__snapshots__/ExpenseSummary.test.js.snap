// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with 1 expense 1`] = `
<div>
  <p>
    Showing 
    1
     
    expense
     totalling 
    $569.40
    .
  </p>
</div>
`;

exports[`should render ExpenseSummary with multiple expenses 1`] = `
<div>
  <p>
    Showing 
    5
     
    expenses
     totalling 
    $54,387,574.98
    .
  </p>
</div>
`;
